app/services/bill-runs/two-part-tariff/determine-licence-issues.service.js

```diff
@@ -220,14 +239,37 @@ function _returnLogIssues (returnLog, licence) {
   return returnLogIssues
 }
 
-function _someReturnsNotReceived (returnLogs, licenceReturnLogs) {
-  const returnLogIds = returnLogs.map((returnLog) => {
-    return returnLog.returnId
-  })
+/**
+ * Determines the received status of returns matched to a charge element.
+ *
+ * - If no matching returns have a status of 'due', `noReturnsReceived` and 'someReturnsNotReceived is `false`.
+ * - If some of the matching returns, but not all, have a status of 'due', `someReturnsNotReceived` is `true`.
+ * - If all matching returns have a status of 'due', `noReturnsReceived` is `true`.
+ */
+function _returnsReceivedStatus (returnLogs, licenceReturnLogs) {
+  const matchingReturnLogs = _getMatchingReturns(returnLogs, licenceReturnLogs)
+
+  let noReturnsReceived
 
-  return licenceReturnLogs.some((licenceReturnLog) => {
-    return returnLogIds.includes(licenceReturnLog.id) && licenceReturnLog.status === 'due'
+  if (matchingReturnLogs.length > 0) {
+    // If no returnsReceived is true, it means every returnLog matched to the charge element has a status of `due`
+    noReturnsReceived = matchingReturnLogs.every((matchingReturnLog) => {
+      return matchingReturnLog.status === 'due'
+    })
+  } else {
+    noReturnsReceived = false
+  }
+
+  // If someReturnsNotReceived is true it means either some of the returnLogs matched to the element have a status of
+  // due, or all of them
+  const someReturnsNotReceived = matchingReturnLogs.some((matchingReturnLog) => {
```

Comment:
I think this could be reworked. if you inverted the test so look for responses that are not equal to 'due' you know for certain that noReturnsReceived is false and can get rid of the if statement above I think? and if you know that at least one of them is not equal to 'due' there's no need to do the every until after that? do you still need the every after that?

Reply:
<img width="1659" alt="Screenshot 2024-07-23 at 17 09 43" src="https://github.com/user-attachments/assets/746289dc-924c-4bfd-a7e2-53ea4e0dffb9">
I managed to get it to work with the `.some` function, but when `some` takes an empty array it returns false, this was being flipped with the bang operator causing the 'noReturnsRecieved' to be true when it shouldn't be. This meant I needed to add the if statement back in, for it to only execute when we know there are returnLogs. As you can see from the photo the only difference between the two working versions is the `.some` or the `.every`. I decided to go with the `every` as I felt it was more readable and you don't have to get your head around any switching on the true/false with the bang operators <img src="https://media0.giphy.com/media/wIhY2p9UtJrUQQw6nz/giphy.gif"/>